pkg/k8s/resource/resource.go

```diff
@@ -532,6 +544,25 @@ func (r *resource[T]) Events(ctx context.Context, opts ...EventsOpt) <-chan Even
 	return out
 }
 
+func (r *resource[T]) resourceName() string {
+	if r.opts.name != "" {
+		return r.opts.name
+	}
+
+	// We create a new pointer to the reconciled resource type.
+	// For example, with resource[*cilium_api_v2.CiliumNode] new(T) returns **cilium_api_v2.CiliumNode
+	// and *new(T) is nil. So we create a new pointer using reflect.New()
+	o := *new(T)
+	sourceObj := reflect.New(reflect.TypeOf(o).Elem()).Interface().(T)
+
+	gvk, err := apiutil.GVKForObject(sourceObj, scheme)
+	if err != nil {
+		return ""
+	}
+
+	return strings.ToLower(gvk.Kind)
```

Comment:
I could be off here but shouldn't the resource name be the pluralize version of kind besides just lower?

Reply:
`GroupVersionKind` doesn't have a pluralized version of kind.
https://github.com/cilium/cilium/blob/6a70af21c3b6985096fca3f8240139b548ac4760/vendor/k8s.io/apimachinery/pkg/runtime/schema/group_version.go#L142-L146

Personally, I don't see the need to make it plural because this function returns the resource name which the workqueue handles. Plus, this approach is the same as the controller-runtime. So the singular form is more consistent.

Reply:
I'm sorry, I missed this before, thanks for catching it  @derailed.

By the k8s API standards, yes, the resource name should _generally_ be the lower-case, pluralized Kind (that is, `Pod` Kind = `pods` resource). But this is a convention, not an invariant rule, so it's not _guaranteed_. (The difference is actually that you can put a _resource_ name directly into a GET URL to the apiserver, and it will work, while a Kind will not.)

But, as @ysksuzuki says, you can't just add an "s" and call it a day. The upstream k8s code will actually do lookups to the apiserver to get the canonical resource name for a GVK - because the apiserver holds mappings for every GVK to resource (so that it can generate URLs itself).

That means, however, that you need a dynamic client, and to use the restmapper package to generate a dynamic restmapper, that you then pass the GVK to, that supplies you with a mapper object that contains the GVR for the given GVK.

I couldn't find old code where I've done this before, but this example, taken from (https://ymmt2005.hatenablog.com/entry/2020/04/14/An_example_of_using_dynamic_client_of_k8s.io/client-go) seems about right:
```go
import (
    "k8s.io/apimachinery/pkg/api/meta"
    "k8s.io/apimachinery/pkg/runtime/schema"
    "k8s.io/client-go/discovery"
    "k8s.io/client-go/discovery/cached/memory"
    "k8s.io/client-go/rest"
    "k8s.io/client-go/restmapper"
)

// find the corresponding GVR (available in *meta.RESTMapping) for gvk
func findGVR(gvk *schema.GroupVersionKind, cfg *rest.Config) (*meta.RESTMapping, error) {

    // DiscoveryClient queries API server about the resources
    dc, err := discovery.NewDiscoveryClientForConfig(cfg)
    if err != nil {
        return nil, err
    }
    mapper := restmapper.NewDeferredDiscoveryRESTMapper(memory.NewMemCacheClient(dc))

    return mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
}
```

That's the only way to _guarantee_ that you've got the correct values completely generically.

_However_, in this case, if we're controlling all the objects this is used for, we _may_ be able to get away with what we have here.

Reply:
@youngnick Thank you for your comment!

To be clear, the `resourceName()` function resolves GVK from a given Go struct type to add `GVK.Kind` to a name label of the workqueue metrics. It's not for finding GVR and requesting kube-apiserver using a dynamic client.　https://github.com/cilium/cilium/pull/27042#issuecomment-1658419028

Regarding the rest API client, the current implementation of `resource[T k8sRuntime.Object]` requires ListWatcher with a typed client to callers. https://github.com/cilium/cilium/blob/9cdcddc67061c59df7b517ef52a1d5a8f0b700c1/pkg/k8s/resource/resource.go#L71-L74

Reply:
🤦 Sorry for the noise then, I guess that's on me for not reading the rest of the code properly and assuming what `resource` meant.